DllImportGenerator/DllImportGenerator/ManualTypeMarshallingHelper.cs

```diff
@@ -0,0 +1,65 @@
+#nullable enable
+
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.Interop
+{
+    static class ManualTypeMarshallingHelper
+    {
+        public const string ValuePropertyName = "Value";
+        public const string StackBufferSizeFieldName = "StackBufferSize";
+        public const string ToManagedMethodName = "ToManaged";
+        public const string FreeNativeMethodName = "FreeNative";
+
+        public static bool HasToManagedMethod(ITypeSymbol nativeType, ITypeSymbol managedType)
+        {
+            return nativeType.GetMembers(ToManagedMethodName)
+                    .OfType<IMethodSymbol>()
+                    .Any(m => m.Parameters.IsEmpty &&
+                        !m.ReturnsByRef &&
+                        !m.ReturnsByRefReadonly &&
+                        SymbolEqualityComparer.Default.Equals(m.ReturnType, managedType) &&
+                        !m.IsStatic);
+        }
+
+        public static bool IsManagedToNativeConstructor(IMethodSymbol ctor, ITypeSymbol managedType)
+        {
+            return ctor.Parameters.Length == 1 &&
+                SymbolEqualityComparer.Default.Equals(managedType, ctor.Parameters[0].Type);
+        }
+
+        public static bool IsStackallocConstructor(
+            IMethodSymbol ctor,
+            ITypeSymbol managedType,
+            ITypeSymbol spanOfByte)
+        {
+            return ctor.Parameters.Length == 2 &&
+                SymbolEqualityComparer.Default.Equals(managedType, ctor.Parameters[0].Type) &&
+                SymbolEqualityComparer.Default.Equals(spanOfByte, ctor.Parameters[1].Type);
+        }
+
+        public static IMethodSymbol? FindGetPinnableReference(ITypeSymbol type)
+        {
+            return type.GetMembers("GetPinnableReference")
```

Comment:
Let's document this `GetPinnableReference()` method discovery. Why and what it means - https://docs.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-7.3/pattern-based-fixed.